Add probe based mechanism for kubelet plugin discovery

vikaschoudhary16 opened the pull request:
**Which issue(s) this PR fixes**
Fixes #56944 
[Design Doc](https://docs.google.com/document/d/1dtHpGY-gPe9sY7zzMGnm8Ywo09zJfNH-E1KEALFV39s/edit#heading=h.7fe6spexljh6)

**Notes For Reviewers**:
Original PR is https://github.com/kubernetes/kubernetes/pull/59963. But because of too many comments(171) that PR does not open sometimes. Therefore this new PR is created to get the github link working.
 
Related PR is https://github.com/kubernetes/kubernetes/pull/58755 
For review efficiency, separating out of the commits or original PR here. 

```release-note
Add probe based mechanism for kubelet plugin discovery
```
/sig node
/area hw-accelerators
/cc @jiayingz @RenaudWasTaken @vishh @ScorpioCPH @sjenning @derekwaynecarr @jeremyeder @lichuqiang @tengqm @saad-ali @chakri-nelluri @ConnorDoyle @vladimirvivien

vikaschoudhary16 committed:
Auto-generated files

vikaschoudhary16 committed:
Add probe based mechanism for kubelet plugin discovery

jiayingz commented:
/lgtm

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

figo commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,243 @@
+		glog.Errorf("Failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	resp := w.invokeRegistrationCallbackAtHandler(infoResp, socketPath)
```
this function has no timeout, potentially can stuck the pluginwatcher.

figo commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,243 @@
+	return &registerapi.RegistrationStatus{PluginRegistered: true}
+}
+
+// Start watches for the creation of plugin sockets at the path
```
this is not about your patch, but about the design of watching fsnotify events, 
this approach is relying on an assumption that: fsnofity will be reliably notify pluginwatcher.
but i am just playing evil here, this assumption maybe not 100% true by looking at number of open issue:
https://github.com/fsnotify/fsnotify/issues?page=3&q=is%3Aissue+is%3Aopen and my local test on Mac also show issue with fsnotify.

the other possible workaround is to traverse folder with interval, how do you think?

RenaudWasTaken replied:
I'm not sure the number of issue is a reliable metric as many issue seem to be questions , feature requests or refer to an old version :)

Additionally I believe fsnotify is production ready given [the number of projects already using it](https://github.com/search?p=1&q=%22github.com%2Ffsnotify%2Ffsnotify%22&type=Code) (including docker and k8s)

vishh replied:
Given the number of plugins we expect is small, traversal might not be a bad choice. However inotify does provide better latency though and avoids unnecessary traversals. @vikaschoudhary16 it would be worth examining the open issues against inotify prior to taking a hard dependency on it.

vikaschoudhary16 replied:
thanks @figo for pointing this out.
@vishh I took a look around this. It seems fsnotify is not being maintained actively. And reading through github issues, looks like people have been maintaining there forks as well for the some of their fixes which could not get merged. Here are the relevant links:
https://github.com/fsnotify/fsnotify/issues/245#issuecomment-388498752

On the other hand, what @RenaudWasTaken said is also true that so many projects are using it. And therefore i think there are chances that this project will likely be able to find a adopter soon. I am slightly in favor of keeping fsnotify, at the same time if majority thinks traversal might be a better option, open to make changes for traversal as well.

@jiayingz @saad-ali @thockin what do you guys think?

vishh replied:
I'd prefer not taking on a dependency that is not being maintained actively. We should either have a plan to contribute back to that project or fork it in k8s organization.

jiayingz replied:
I would prefer to staying with fsnotify for now. Mostly we just need a go pkg for inotify syscall. The fsnotify pkg is already included in k8s so we are not adding a new dependency, and so far we haven't heard issues in the past for the included version. It will be a much easier change to switch to a more stable fsnotify go pkg in the future than building on top of a different model where the latency to detect a new plugin would be affected by how long and how often directory traverse happens.

figo reviewed (commented):
overall looks good, i have been trying to understand every line of code in this change in the past two days, i have two comments as below, thanks for the contribution.

figo commented on pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go:
```diff
@@ -0,0 +1,150 @@
+		require.Equal(t, PluginName, name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		_, conn, err := dial(socketPath)
```
should be 'sockPath' instead of 'socketPath'

tallclair commented on pkg/kubelet/util/pluginwatcher/README:
```diff
@@ -0,0 +1,29 @@
+
+Here are the general rules that Kubelet plugin developers should follow:
+- Run as priviledged. Currently creating socket under PluginsSockDir requires
+  priviledged access.
```
I don't think that's true? Doesn't that just require running as root & a HostPath mount? I think it's likely that plugins will still need to be privileged, but for different reasons...

vishh replied:
+1 If all that's needed is access to a root owned directory, that could come in the form of running as gid `0` too right?

tallclair replied:
Depending on the access permissions, but I'd recommend requiring a root uid (i.e. `700` permissions). The k8s controls limit access to gid 0 are still alpha.

tallclair commented on pkg/kubelet/util/pluginwatcher/README:
```diff
@@ -0,0 +1,29 @@
+- Implements the Registration service specified in
+  pkg/kubelet/apis/pluginregistration/v*/api.proto.
+- The plugin name sent during Registration.GetInfo grpc should be unique
+  for the given plugin type (CSIPlugin or DevicePlugin).
```
Are there guidelines around namespacing? E.g. "mycompany.com/my-plugin"

vishh replied:
This is a list of plugin APIs that kubelet natively supports whose cardinality is very limited. The names of individual plugins on the other hand needs to adhere to existing namespacing norms.

vikaschoudhary16 replied:
> Are there guidelines around namespacing? E.g. "mycompany.com/my-plugin"

Yes, it should be unique because socket name is expected to have one-to-one mapping with plugin name and it should not be hierarchical. Both points are mentioned.

tallclair commented on pkg/kubelet/util/pluginwatcher/README:
```diff
@@ -0,0 +1,29 @@
+- The socket path needs to be unique and doesn't conflict with the path chosen
+  by any other potential plugins. Currently we only support flat fs namespace
+  under PluginsSockDir but will soon support recursive inotify watch for
+  hierarchical socket paths.
```
`com.mycompany.my-plugin`?

vladimirvivien replied:
If the PluginsSockDir is flat, I think you should at least have PluginsSockDir/csi and PluginsSockDir/devices with two watcher instances.  That way you can avoid name collisions between csi and device plugin names.

RenaudWasTaken commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,243 @@
+	defer cancel()
+	infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
+	if err != nil {
+		glog.Errorf("Failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
```
Either return an error or log the error

RenaudWasTaken commented on pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go:
```diff
@@ -0,0 +1,150 @@
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		_, conn, err := dial(socketPath)
+		require.Nil(t, err)
```
`require.NoError` ?

vishh commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,243 @@
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierarchical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			w.registerPlugin(path.Join(w.path, f.Name()))
```
Why do you think this will involve significant latency?

RenaudWasTaken replied:
This call is blocking with a timeout of 10s which means the worst case is n * 10s (n number of plugins in the directory).

vikaschoudhary16 replied:
i would prefer the first one, making call goroutine at L#86. If we pass create events, would not they all be processed serially at L#182? 
Or another option could be pass create events and then call `registerPlugin` in a goroutine at L#182. 
I still prefer first one because of simplicity. Thoughts?

RenaudWasTaken replied:
> i would prefer the first one, making call goroutine at L#86. If we pass create events, would not they all be processed serially at L#182?

Isn't that already the case for all other plugins?

vikaschoudhary16 replied:
but that is not related to kubelet start time. Kubelet start time will be impacted by only those plugins, sockets of which are already there and if we register them serially. In the updated patch, `registerPlugin` is being invoked in a goroutine. PTAL

RenaudWasTaken replied:
This makes no sense, what the original discussion was about is doing registration outside the main thread.

Whether we choose to do registration of multiple plugins serially or in parallel should not impact the Kubelet start time.

I don't have a preference for the option you choose. I do however think that registration of multiple plugin should follow the same pattern at startup or during runtime.

- Option 1 (sending an event in the event channel) has the nice property of having a single point of entry for both step of the lifecycle thus making refactoring easier.
- Option 2 (have registration of multiple plugins in parallel ) also seems like a sensible approach. 

Additionally it seems like you could even combine both approach.

vladimirvivien commented on pkg/kubelet/util/pluginwatcher/README:
```diff
@@ -0,0 +1,29 @@
+
+Here are the general rules that Kubelet plugin developers should follow:
+- Run as priviledged. Currently creating socket under PluginsSockDir requires
+  priviledged access.
```
Further down, someone mentions namespacing the plugin names.. another place namespacing may make sense is under the `PluginsSockDir`.  Maybe separate plugins socket path based on type i.e. `$PluginsSockDir/csi/<plugins-name>` or `$PluginsSockDir/devices/<plugins-name>` etc.

vladimirvivien commented on pkg/kubelet/util/pluginwatcher/README:
```diff
@@ -0,0 +1,29 @@
+  under PluginsSockDir but will soon support recursive inotify watch for
+  hierarchical socket paths.
+- A plugin should clean up its own socket upon exiting or when a new instance
+  comes up. A plugin should NOT remove any sockets belonging to other plugins.
```
Is there anyway to prevent a plugin from removing sockets dir from another plugin ?  This could be a big issue.

vikaschoudhary16 replied:
I dont have a approach for this on top of my mind. Would love to hear suggestions.

vishh replied:
One option would be that of using a directory per plugin where a plugin can be restricted to only create sockets within a specific directory.
Unless plugins stop running as privileged or uid `0`, attempts to secure plugins from one-another would be moot.

vladimirvivien commented on pkg/kubelet/util/pluginwatcher/README:
```diff
@@ -0,0 +1,29 @@
+different types of node-level plugins such as device plugins or CSI plugins.
+It discovers plugins by monitoring inotify events under the directory returned by
+kubelet.getPluginsDir(). Lets refer this directory as PluginsSockDir.
+For any discovered plugin, pluginwatcher issues Registration.GetInfo grpc call
```
Is it too late to turn watcher into a service? The plugin calls the watcher at a prescribed UDS location to register itself instead of the watcher doing inotify on a flat directory.  That way the watcher can apply stricter registration rules including voiding name clashes, etc.

vikaschoudhary16 replied:
If i understood correctly, you are suggesting model 1, where plugin starts the communication. As explained the design doc, there is a problem in re-registration case with model 1. Therefore we decided to implement model 2. 
flat directory is just for initial implementation and will be extended to hierarchical in successive PR. 
I think in model 1 also onus of avoiding plugin name conflict is on plugins. I am not sure how model 1 is better than model 2 in avoiding name clashes.

vladimirvivien replied:
@vikaschoudhary16 I am concerned with plugin name clash between plugin types of CSI and DevicePlugin plugins.  If PluginsSockDir is flat then a plugin name from CSI may clash with a plugin from DevicePlugin.

vladimirvivien reviewed (commented):
Apologies for making design suggestions so late.

figo commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,243 @@
+func (w *Watcher) invokeRegistrationCallbackAtHandler(infoResp *registerapi.PluginInfo, socketPath string) *registerapi.RegistrationStatus {
+	var handlerCbkFn RegisterCallbackFn
+	var ok bool
+	handlerCbkFn, ok = w.handlers[infoResp.Type]
```
should check the plugin type? which defined at apis/pluginregistration/v1alpha1/constants.go

RenaudWasTaken replied:
I think this is what the next line does unless I didn't understand your comment :) ?

figo replied:
no, if you look at the example_plugin.go, it using plugin type = "example-plugin-type",
basically registerapi.PluginInfo.Type is not been enforced

RenaudWasTaken replied:
Do you mean strictly enforce the plugins in constants.go? In that case I'm not sure that's needed.

Plugin type is enforced by the fact that the type is or isn't in the array. Adding to the array is a matter of registering a plugin, in the case of the test, it passes because a handler added with that type.

figo replied:
right, i mean the first part, otherwise i don't see the purpose of defining plugin type const at constants.go

vikaschoudhary16 replied:
can you please explain more on security worry part? `AddHandler` can only be invoked in a static manner from within the kubelet code.

figo replied:
sorry guys, security is not a concern after some thought, the only improvement i can think of is: add log even when handler been invoked successfully at invokeRegistrationCallbackAtHandler(), this is for audit purpose base on AAA model,  this logging improvement can be applied on other functions as well.

RenaudWasTaken commented on pkg/kubelet/util/pluginwatcher/example_plugin.go:
```diff
@@ -0,0 +1,148 @@
+		defer e.wg.Done()
+		// Blocking call to accept incoming connections.
+		err := e.grpcServer.Serve(lis)
+		glog.Errorf("example server stopped serving: %v", err)
```
I was just mentioning that you can add a check to see if the error is different than nil before printing it.

vikaschoudhary16 commented:
@tallclair @jiayingz @figo @RenaudWasTaken @vishh @vladimirvivien 
Addressed most of the comments. PTAL
@vishh regarding `Info/GetInfo` started a discussion on RMWG slack. Trying to avoid the unicorn again because of too  many comments on the PR.

RenaudWasTaken reviewed (commented):
The `registrationStatus` race still seems wrong to me.

I'd expect the lifecycle of a plugin to always receive a synchronous `registrationStatus` call before any other step is performed.

Basically the current model looks like the following:
```golang
go devicePluginClient.ListWatch(ctx, ...)
go pluginWatcherClient.NotifyRegistrationStatus(ctx, resp)
```

Whereas I'd expect it to look more like this:
```golang
pluginWatcherClient.NotifyRegistrationStatus(ctx, resp)
devicePluginClient.ListWatch(ctx, ...)
```

vladimirvivien commented:
@vikaschoudhary16 any thought on getting this approved this quarter ? We're ready to give it a try in CSI.

sbezverk commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,240 @@
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
```
It would be great to be able to not call the driver directly, because it kind of makes it a kubernetes specific and one of goals of CSI for the driver to be completely CO agnostic, but instead to use an interface, which can be generic f definition but underneath with CSI or another plugin specific implementation.  Example, to be able to reply to GetInfo call, all CSI drivers will have to be aware of this type of request which is kind of kubernetes specific and it goes a bit against CSI approach. If registerPlugin is implemented as interface, and then each plugin uses plugin specific way to get required information, then this code is kept generic but we do not loose flexibility.
@saad-ali @vladimirvivien Would you agree  with proposed modification?

jiayingz replied:
I thought CSI sidecar container will handle the interaction with kubelet, which would also include supporting this registerapi. Can some CSI folks double verify this?

sbezverk replied:
@jiayingz If side car handles `GetInfo` request then it would require two sockets, one opened by CSI driver for CSI related communication and one opened by sidecar to reply to GetInfo request,  how do you plan to deal now with two opened sockets? Where would sidecar create `GetInfo` socket?

sbezverk replied:
Defining registration interface, imho would be the cleanest solution.

sbezverk replied:
After discussion within CSI WG, please disregards these comments for now. In the interest of time, this PR should be merged and revisited later if required.

vikaschoudhary16 commented:
@vladimirvivien 
> @vikaschoudhary16 any thought on getting this approved this quarter ? We're ready to give it a try in CSI.

I would like to have that and targeting for the same.

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-kops-aws

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-kops-aws

vladimirvivien commented:
/lgtm

vladimirvivien commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vladimirvivien commented:
/cancel lgtm

vladimirvivien commented:
Hi @vikaschoudhary16 , apologies for the last minute design changes.  After couple of CSI review cessions regarding implementing this PR, the CSI team wants to suggest the following changes which will work better for the CSI plugin model.  They are minor and hope will work for you as well.

1)  Add field `endpoint` to the `PluginInfo` message
```protobuf
message PluginInfo {
	// Type of the Plugin. CSIPlugin or DevicePlugin
	string type = 1;

	// Plugin name that uniquely identifies the plugin for the given plugin type.
	// For DevicePlugin, this is the resource name that the plugin manages and
	// should follow the extended resource name convention.
	// For CSI, this is the CSI driver name.
	string name = 2;
    
        // The endpoint location where the plugin will be listening for incoming requests.
        // This allows the plugin to register using one endpoint and possibly use
        // a different socket for control operations.  CSI uses this model to delegate
        // its registration external from the plugin.
        string endpoint;

	// Plugin service API versions the plugin supports.
	// For DevicePlugin, this maps to the deviceplugin API versions the
	// plugin supports at the given socket.
	// The Kubelet component communicating with the plugin should be able
	// to choose any preferred version from this list, or returns an error
	// if none of the listed versions is supported.
	repeated string supported_versions = 3;
}
```
2) Change the callback function definition to take the entire `PluginInfo`
```go
type RegisterCallbackFn func(*registerapi.PluginInfo) (error, chan bool)
```
See code here : https://github.com/vikaschoudhary16/kubernetes/blob/5eeb4e995e194a1cd290026437373cfe86875f8f/pkg/kubelet/util/pluginwatcher/plugin_watcher.go#L37

**Or**
If you don't want to pass in the entire `PluginInfo` protobuf value to the callback func (as defined above), change the following
code:
https://github.com/vikaschoudhary16/kubernetes/blob/5eeb4e995e194a1cd290026437373cfe86875f8f/pkg/kubelet/util/pluginwatcher/plugin_watcher.go#L144

to:
```go
err, chanForAckOfNotification := handlerCbkFn(infoResp.Name, versions, infoResp.Endpoint)
```

vikaschoudhary16 commented:
@vladimirvivien Effort required is not much to add this change but i just want to make sure if it is really needed. I am just comparing your suggestion against the approach where CSI manager will query this additional control socket over the endpoint provided by pluginwatcher. Are there any challenges in using wathcer provided socket to query control socket?

@jiayingz @RenaudWasTaken @vishh @saad-ali would you like to weigh in?

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vladimirvivien commented:
@vikaschoudhary16 it is needed because of following reasons:

Because CSI is industry wide, the external CSI storage drivers are implemented **to be as non-K8s-specific as possible**.  To do this, CSI drivers use a secondary binary (Registrar sidecar) to handle k8s-specific deployment functions.  That binary will implement the Kubelet Plugin `Registration` service for plugin registration.  When the watcher queries the Registrar sidecar, it then provides the actual socket for the external CSI driver.  

So, for  CSI, the registration flow is as follows: 

Registrar sidecar provides socket ➡️  Watcher detects socket ➡️  Watcher calls `GetInfo` against Registrar ➡️  Registrar returns `PluginInfo` including `PluginInfo.Endpoint` ➡️  Watcher calls `InternalCSIPluginMgr.CallBackFn(PluginInfo.EndPoint)`

At this point, the registration is complete and now the internal CSI manager knows can communicate with the external CSI storage driver using the endpoint info it received in the last step above.

This also allows the future plugin framework designers to separate internal plugin socket structure from that dictated by the kubelet plugin architecture.  That way, plugin framework designers can organize their socket directory as needed and not forced to use one prescribed by kubelet plugin. 

Hope this clarifies my intentions for this design change.

vladimirvivien commented:
@vikaschoudhary16 I will do a retest before lgtm.

/retest

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vladimirvivien commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-local-e2e-containerized

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-bazel-test
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-kubemark-e2e-gce-big

vladimirvivien commented:
/lgtm

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

jiayingz commented:
/lgtm

vishh reviewed (approved):
/approve
/lgtm

derekwaynecarr commented:
/status approved-for-milestone

vikaschoudhary16 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-kubemark-e2e-gce-big

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-kops-aws

RenaudWasTaken commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -0,0 +1,260 @@
+)
+
+// RegisterCallbackFn is the type of the callback function that handlers will provide
+type RegisterCallbackFn func(pluginName string, endpoint string, versions []string, socketPath string) (error, chan bool)
```
An interface model might be more explicit? i.e:
```golang
type RegisterCallback interface {
    Validate(...)
    Register(...)
}
```

RenaudWasTaken reviewed (commented):
/lgtm

We can discuss channel vs interface in another PR / issue / on slack, I'll take care of it :)
